cmd/snap/cmd_login.go

```diff
@@ -54,28 +53,33 @@ func init() {
 }
 
 func requestLoginWith2faRetry(username, password string) error {
+	var otp []byte
+	var err error
+
+	var msgs = [3]string{
+		i18n.G("Two-factor code: "),
+		i18n.G("Bad code. Try again: "),
+		i18n.G("Wrong again. Last chance: "),
+	}
+
 	cli := Client()
-	// first try without otp
-	_, err := cli.Login(username, password, "")
-	if err != nil {
-		// check if we need 2fa
-		if err, ok := err.(*client.Error); !ok || err.Kind != store.TwoFactorErrKind {
+	reader := bufio.NewReader(nil)
+
+	for i := 0; ; i++ {
+		// first try without otp
+		_, err = cli.Login(username, password, string(otp))
+		if i > 2 || !client.IsTwoFactorError(err) {
 			return err
 		}
 
-		fmt.Print(i18n.G("Two-factor code: "))
-		reader := bufio.NewReader(os.Stdin)
+		reader.Reset(os.Stdin)
+		fmt.Print(msgs[i])
```

Comment:
Should this use the package-level Stdout, so it's redirected for testing?

I know we have a few tests which are currently missing this and `go test` ends up breaking a few lines up already on test runs.